src/cpu/flexcpu/flexcpu.cc

```diff
@@ -0,0 +1,1103 @@
+
+    memoryUnit.schedule();
+
+    return true;
+}
+
+void
+FlexCPU::startup()
+{
+    BaseCPU::startup();
+
+    for (auto& thread : threads) {
+        thread->startup();
+    }
+    // TODO add code which needs to be run during the first simulation event
+}
+
+void
+FlexCPU::suspendContext(ThreadID tid)
+{
+    // Need to de-schedule any instructions in the pipeline?
+    BaseCPU::suspendContext(tid);
+}
+
+void
+FlexCPU::haltContext(ThreadID tid)
+{
+    // Need to de-schedule any instructions in the pipeline?
+    BaseCPU::haltContext(tid);
+}
+
+void
+FlexCPU::switchOut()
+{
+    // TODO flush speculative state.
+    // TODO clear any outstanding memory requests.
+
+    BaseCPU::switchOut();
+}
+
+void
+FlexCPU::takeOverFrom(BaseCPU* cpu)
+{
+    // TODO may want to ensure clean speculative state prior to attempting?
+
+    BaseCPU::takeOverFrom(cpu); // BaseCPU::takeOverFrom() already handles the
+                                // calls to ThreadContext::takeOverFrom() via
+                                // the internal ThreadContext* table.
+
+    // TODO load any other state.
+}
+
+Counter
+FlexCPU::totalInsts() const
+{
+    Counter insts = 0;
+    for (auto &thread : threads) {
+        insts += thread->getNumInsts();
+    }
+    return insts;
+}
+
+Counter
+FlexCPU::totalOps() const
+{
+    Counter ops = 0;
+    for (auto &thread : threads) {
+        ops += thread->getNumOps();
+    }
+    return ops;
+}
+
+void
+FlexCPU::wakeup(ThreadID tid)
+{
+    assert(tid < numThreads);
+
+    // Get the ThreadContext interface pointer from BaseCPU::getContext(), for
+    // flexibility in how the ThreadContext is retrieved (in case we change
+    // from using SimpleThread).
+    ThreadContext* tc = getContext(tid);
+    if (tc->status() == ThreadContext::Suspended) {
+        tc->activate();
+    }
+}
+
+bool
+FlexCPU::DataPort::recvTimingResp(PacketPtr pkt)
+{
+    panic_if(!cpu->outstandingMemReqs.count(pkt),
+        "Received a data port response we don't remember sending!");
+
+    auto& callback = cpu->outstandingMemReqs[pkt];
+
+    callback();
+
+    cpu->outstandingMemReqs.erase(pkt);
+    delete pkt;
+
+    return true;
+}
+
+void
+FlexCPU::DataPort::recvReqRetry()
+{
+    assert(blockedPkt);
+    DPRINTF(FlexCPUCoreEvent, "data retrying %s\n", blockedPkt->print());
+
+    PacketPtr pkt = blockedPkt;
+    blockedPkt = nullptr;
+    sendPacket(pkt);
+
+    if (!blockedPkt) {
+        DPRINTF(FlexCPUCoreEvent, "Sent successfully!\n");
+        // If we're now unblocked, try to schedule other things from memoryunit
+        cpu->memoryUnit.schedule();
+    }
+}
+
+void
+FlexCPU::DataPort::sendPacket(PacketPtr pkt)
+{
+    DPRINTF(FlexCPUCoreEvent, "Data sending %s\n", pkt->print());
+    assert(!blockedPkt);
+    if (!sendTimingReq(pkt)) {
+        DPRINTF(FlexCPUCoreEvent, "Data failed to send %s\n",
+                 pkt->print());
+        blockedPkt = pkt;
+    }
+}
+
+bool
+FlexCPU::InstPort::recvTimingResp(PacketPtr pkt)
```

Comment:
Can we just have two instances of Port?

Reply:
Probably possible, but probably not worth it.